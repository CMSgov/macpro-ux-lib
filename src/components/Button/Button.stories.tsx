import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "./Button";

export default {
  title: "USWDS/Base/Button",
  component: Button,
  argTypes: {
    buttonVariation: {
      defaultValue: "primary",
      description: "This is the style prop for the button",
    },
    buttonText: {
      defaultValue: "Button",
      description: "This is the text prop for the button",
    },
    disabled: {
      defaultValue: false,
      control: "boolean",
      description: "Determines whether or not a button is enabled",
    },
    icon: {
      description: "Handles the icon to render in the button",
    },
    shiftIconLeft: {
      description:
        "Determines if a rendered icon should render on the left side",
    },
    largeButton: {
      description: "Determines whether a large button is required",
    },
    ariaLabel: {
      description:
        "Overwrites the aria label of the button element that will be read on screen readers",
    },
    onClick: {
      description: "Handles its behavior when the button is clicked",
      action: "clicked",
    },
    type: {
      defaultValue: "button",
      options: ["submit", "reset", "button"],
      control: { type: "select" },
      description: "Type of specified button",
    },
    target: {
      defaultValue: "_self",
      options: ["_blank", "_self", "_parent", "_top"],
      control: { type: "select" },
      description:
        "Specifies a name or a keyword that indicates where to display the response that is received after clicking the button",
    },
  },
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = ({ buttonText, ...rest }) => (
  <Button buttonText={buttonText} {...rest} />
);

export const PrimaryButton = Template.bind({});
PrimaryButton.args = {
  buttonVariation: "primary",
};

export const SecondaryButton = Template.bind({});
SecondaryButton.args = { buttonVariation: "secondary" };
export const ErrorButton = Template.bind({});
ErrorButton.args = { buttonVariation: "error" };
export const SuccessButton = Template.bind({});
SuccessButton.args = { buttonVariation: "success" };
export const InverseButton = Template.bind({});
InverseButton.args = { buttonVariation: "inverse" };
export const BaseButton = Template.bind({});
BaseButton.args = { buttonVariation: "base" };
export const LinkButton = Template.bind({});
LinkButton.args = { buttonVariation: "link" };
export const BigButton = Template.bind({});
BigButton.args = { largeButton: true };
export const OutlineButton = Template.bind({});
OutlineButton.args = { buttonVariation: "outline" };
