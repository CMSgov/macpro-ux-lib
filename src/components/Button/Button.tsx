import React from "react";

type IntrinsicElements = JSX.IntrinsicElements["button"];

interface Props extends IntrinsicElements {
  buttonText: string;
  buttonVariation?: ButtonVariation;
  disabled?: boolean;
  icon?: boolean;
  shiftIconLeft?: boolean;
  ariaLabel?: string;
  largeButton?: boolean;
}

/**
 * Button Component
 * @param {string}  buttonText            Renders the text contained in the button.
 * @param {string}  [buttonVariation]     Renders the style of the button.
 * @param {boolean} [disabled]            Determines whether or not a button is enabled.
 * @param {Icon}    [icon]                Handles the icon to render in the button.
 * @param {boolean} [shiftIconLeft]       Determines if a rendered icon should render on the left side.
 * @param {boolean} [ariaLabel]           Overwrites the aria label of the button element that will be read on screen readers.
 * @param {boolean} [largeButton]         Determines whether a large button is required.
 * @param {string}  [button]              Type of specified button.
 * @param {Event}   [onClick]            Handles its behavior when the button is clicked.
 * @param {string}  [target]              Specifies a name or a keyword that indicates where to display the response that is received after clicking the button.
 */

type ButtonVariation =
  | "primary"
  | "secondary"
  | "error"
  | "success"
  | "inverse"
  | "base"
  | "link"
  | "outline";

const ButtonVariationConversion: { [key: string]: string } = {
  primary: "",
  secondary: "outline",
  error: "secondary",
  success: " bg-green",
  inverse: "outline text-gray-70 bg-base-lighter",
  base: "base",
  link: "unstyled",
  outline: "outline text-black",
};

export const Button: React.FC<Props> = ({
  buttonVariation = "primary",
  buttonText,
  shiftIconLeft,
  icon,
  ariaLabel,
  largeButton = false,
  ...rest
}) => {
  const buttonVariationType =
    ButtonVariationConversion[buttonVariation || ""] ?? "";
  return (
    <button
      {...rest}
      aria-label={ariaLabel || `${buttonText} button`}
      className={`usa-button usa-button--${buttonVariationType} ${
        largeButton ? "usa-button--big" : ""
      }`}
    >
      {shiftIconLeft && icon && <span className="margin-right-1">+</span>}
      <span>{buttonText}</span>
      {icon && !shiftIconLeft && <span className="margin-left-1">+</span>}
    </button>
  );
};
