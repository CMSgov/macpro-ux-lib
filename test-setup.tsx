import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
export const cleanAttributes = (container: HTMLElement, prefixes: string[]) => {
  const attributes = ["id", "for"];

  const clean = (attr: string) => {
    for (const prefix of prefixes) {
      if (attr.includes(prefix)) {
        const attrList = attr.split("-");
        attrList.splice(-1);
        return attrList.join("-");
      }
    }
    return attr;
  };

  attributes.forEach((attr) => {
    if (container.hasAttribute(attr)) {
      const value = clean(container.getAttribute(attr)!);
      container.setAttribute(attr, value);
    }
  });

  Array.from(container.children).forEach((child) => {
    cleanAttributes(child as HTMLElement, prefixes);
  });
};
