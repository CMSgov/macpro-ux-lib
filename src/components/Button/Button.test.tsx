import React from "react";
import { Button } from "./Button";
import { fireEvent, screen, render } from "../../test-setup";

describe("Tests for the button component.", () => {
  const mockChangeFn = jest.fn();

  it("Should render and fire a callback", () => {
    render(
      <Button data-testid="button" buttonText="Button" onClick={mockChangeFn} />
    );
    const buttonComp = screen.getByTestId("button");

    fireEvent.click(buttonComp);
    expect(buttonComp).toBeInTheDocument();
    expect(buttonComp).toHaveTextContent("Button");
    expect(mockChangeFn).toHaveBeenCalled();
  });

  it("Should render a disabled button with an aria label of disabled big button", () => {
    render(
      <Button
        data-testid="button"
        disabled
        buttonText="Disabled Button"
        onClick={mockChangeFn}
        ariaLabel="disabled big button"
        largeButton
        buttonVariation="inverse"
      />
    );
    const buttonComp = screen.getByTestId("button");

    expect(buttonComp).toBeInTheDocument();
    expect(buttonComp).toHaveTextContent("Disabled Button");
    expect(buttonComp).toBeDisabled();
    expect(buttonComp).toHaveAttribute("aria-label", "disabled big button");
  });

  describe("compontent snapshots", () => {
    it("primary big disabled", () => {
      const { container } = render(
        <Button
          data-testid="button"
          buttonText="Big Inverse Button"
          onClick={mockChangeFn}
          ariaLabel="big inverse button"
          largeButton
          buttonVariation="inverse"
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("secondary button", () => {
      const { container } = render(
        <Button
          data-testid="button"
          buttonText="Submit"
          onClick={mockChangeFn}
          ariaLabel="Secondary button"
          buttonVariation="secondary"
          iconName=""
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
