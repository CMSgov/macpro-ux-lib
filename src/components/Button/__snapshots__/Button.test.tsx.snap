// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the button component. compontent snapshots primary big disabled 1`] = `
<div>
  <button
    aria-label="big inverse button"
    class="usa-button usa-button--outline text-gray-70 bg-base-lighter usa-button--big"
    data-testid="button"
  >
    <span>
      Big Inverse Button
    </span>
  </button>
</div>
`;

exports[`Tests for the button component. compontent snapshots secondary button 1`] = `
<div>
  <button
    aria-label="Secondary button"
    class="usa-button usa-button--outline "
    data-testid="button"
  >
    <span>
      Button
    </span>
  </button>
</div>
`;
